src/main/scala/com.snowplowanalytics.sauna/utils/Command.scala

```diff
@@ -0,0 +1,224 @@
+      case Success(key) => JsSuccess(key)
+      case Failure(error) => JsError(error.toString)
+    }
+    case _ => JsError("Non-string SchemaKey")
+  }
+
+  implicit val selfDescribingDataReads: Reads[SelfDescribing] = (
+    (JsPath \ "schema").read[SchemaKey] and
+      (JsPath \ "data").read[JsValue]
+    ) (SelfDescribing.apply _)
+
+  implicit val saunaCommandReads: Reads[SaunaCommand] = (
+    (JsPath \ "envelope").read[SelfDescribing] and
+      (JsPath \ "command").read[SelfDescribing]
+    ) (SaunaCommand.apply _)
+
+  sealed trait Semantics
+
+  case object AT_LEAST_ONCE extends Semantics
+
+  implicit val semanticsReads: Reads[Semantics] = Reads {
+    case JsString(s) => s match {
+      case "AT_LEAST_ONCE" => JsSuccess(AT_LEAST_ONCE)
+      case _ => JsError(s"Invalid Semantics: $s")
+    }
+    case _ => JsError("Non-string Semantics")
+  }
+
+  /**
+   * Parameters specifying a command's execution.
+   *
+   * @param semantics  Controls how the command should be executed; currently the
+   *                   only option here is AT_LEAST_ONCE, meaning that the command
+   *                   will be actioned at least one time (but potentially more)
+   * @param timeToLive The command's lifetime (in ms): if a command is
+   *                   received after `timeToLive` ms from its' creation have
+   *                   passed, it will be discarded. (If None, a command won't be
+   *                   discarded based on creation time.)
+   */
+  case class ExecutionParams(
+    semantics: Semantics,
+    timeToLive: Option[Int]
+  )
+
+  implicit val timeToLiveReads: Reads[Option[Int]] = JsPath.readNullable[Int]
+  implicit val executionParamsReads: Reads[ExecutionParams] = Json.reads[ExecutionParams]
+
+  /**
+   * A Sauna command envelope.
+   *
+   * @param commandId   A unique identifier (uuid4) for the command.
+   * @param whenCreated The command's creation date.
+   * @param execution   Parameters specifying a command's execution.
+   * @param tags        A way of attaching additional freeform metadata to a given command.
+   */
+  case class CommandEnvelope(
+    commandId: String,
+    whenCreated: LocalDateTime,
+    execution: ExecutionParams,
+    tags: Map[String, String]
+  )
+
+  implicit val commandEnvelopeReads: Reads[CommandEnvelope] = Json.reads[CommandEnvelope]
+
+  /**
+   * Attempts to extract a Sauna command from a [[JsValue]]. If successful,
+   * passes it on to the processing method.
+   *
+   * @param json         The [[JsValue]] to extract a command from.
+   * @param igluResolver Optional [[Resolver]] containing info about Iglu repositories.
+   * @tparam T The type of the command's data.
+   * @return Right containing a tuple of the command's envelope and data
+   *         if the extraction and processing was successful, Left containing
+   *         an error message otherwise.
+   */
+  def extractCommand[T](json: JsValue,
+    igluResolver: Resolver = Resolver(500, List(
```

Comment:
Also, if we're making Iglu Central resolver special (this should be changed in next release) - could you please instantiate it somewhere once:

```scala
lazy val igluCentralResolver = Resolver(500, List(HttpRepositoryRef(RepositoryRefConfig("Iglu central", 0, List("com.snowplowanalytics")), "http://iglucentral.com")))
```

It contains internal cache and it quite wasteful to instantiate them for each function call.